Swagger Spec for Microsoft.ADHybridHealthService

gamitra opened the pull request:
This Pull Request contains the Swagger specification for Microsoft.ADHybridHealthService RP.
This RP is for Azure Active Directory Connect Health service, which has been GA-ed for last 2 years.
Azure Active Directory (Azure AD) Connect Health helps you monitor and gain insights into your on-premises identity infrastructure and the synchronization services. It enables you to maintain a reliable connection to Office 365 and Microsoft Online Services by providing monitoring capabilities for your key identity components such as Active Directory Federation Services (AD FS) servers, Azure AD Connect servers (also known as Sync Engine), Active Directory domain controllers, etc. It also makes the key data points about these components easily accessible so that you can get usage and other important insights to make informed decisions.
The information is presented in the Azure AD Connect Health portal. In the Azure AD Connect Health portal, you can view alerts, performance monitoring, usage analytics, and other information. Azure AD Connect Health enables the single lens of health for your key identity components in one place.
More details about this RP can be found at:
https://docs.microsoft.com/en-us/azure/active-directory/connect-health/active-directory-aadconnect-health

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

gamitra committed:
Swagger Spec for Microsoft.ADHybridHealthService

This Pull Request contains the Swagger specification for Microsoft.ADHybridHealthService RP.
This RP is for Azure Active Directory Connect Health service, which has been GA-ed for last 2 years.
Azure Active Directory (Azure AD) Connect Health helps you monitor and gain insights into your on-premises identity infrastructure and the synchronization services. It enables you to maintain a reliable connection to Office 365 and Microsoft Online Services by providing monitoring capabilities for your key identity components such as Active Directory Federation Services (AD FS) servers, Azure AD Connect servers (also known as Sync Engine), Active Directory domain controllers, etc. It also makes the key data points about these components easily accessible so that you can get usage and other important insights to make informed decisions.
The information is presented in the Azure AD Connect Health portal. In the Azure AD Connect Health portal, you can view alerts, performance monitoring, usage analytics, and other information. Azure AD Connect Health enables the single lens of health for your key identity components in one place.
More details about this RP can be found at:
https://docs.microsoft.com/en-us/azure/active-directory/connect-health/active-directory-aadconnect-health

gamitra committed:
Update ADHybridHealthService.json

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-libraries-for-java

AutorestCI commented:
# Automation for azure-sdk-for-node

annatisch commented:
@gamitra - please add a readme file for this spec into the resource-manager directory.
You can see here for an example:
https://github.com/Azure/azure-rest-api-specs/tree/master/specification/cognitiveservices/resource-manager

Please also include the sections and/or files for the various language SDK generation.

gamitra committed:
Adding readme file

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5420 @@
+          }
+        },
+        "operationId": "services_GetServices",
+        "parameters": [
```
This is actually a list operation - so should probably be renamed accordingly and use the x-ms-pageable extension:
https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-pageable

annatisch commented:
@gamitra 
Before I go any further with the review - there's a couple of larger things across the spec that need a look at.
- It looks like many of the operation ID's need to be tweaked. We generally don't repeat the noun when that operation is an action on the operation group, e.g.:
services_getService => service_get
services_deleteService => service_delete
- It looks like "serviceMember" needs to be it's own operation group:
"serviceMember_getDataFreshness"
"serviceMember_getExportStatus"
"serviceMember_getMetrics"
- Take a look at the [paging extension](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-pageable) for array response bodies. There are a couple of areas where it could be applied (anything with a single array in the response or anything the uses a "nextlink" field). This ensures that the SDK generators will automatically handle a next link returned by the service, and wrap the response according to each language.
- Avoid using 'post' in an operation name - replace with a more human friendly verb. Use of 'get' and 'patch' is generally okay though.
- Some capitalization could be looked at, as this affects language specific code generation e.g.:
forestsummary -> forestSummary
metricmetadata -> metricMetadata
replicationstatus -> replicationStatus

gamitra commented:
@annatisch , Thank You for these great suggestions. I am modifying the json file using your suggestion. Below are the formats I am using:
1.  For routes containing addsservices, the naming convention for operationId: addsservice_get/list/update/add* . I cant combine them with /servicemembers because they are dedicated to a specific set of Service monitoring(Active Directory Domain Service.)
2. For routes containing /services, I am using the naming convention services_get/list/patch/add*, etc.
3. For routes containing /servicemembers( which are not a part of addsservices route), I am using the naming convention servicemembers_get/ad/patch/*
4. I am adding x-ms-pageable, where the response has a nextLink.

Let me know if this looks fine to you.

annatisch commented:
Thanks @gamitra - looks good.
One thing I would add is that the response does not need a nextlink in order to use x-ms-pageable. This should be used anywhere where the response is a list (then in the spec you can just set `"nextLinkName"` to `null`)

gamitra committed:
Uploading Example files by renaming Post* to Add*

gamitra committed:
Delete PostAlertFeedback

gamitra committed:
Delete PostService.json

gamitra committed:
Delete PostServiceMembers

gamitra committed:
Updating ADhybridhealthService.json with feedback comments

gamitra commented:
@annatisch , missed your previous comment regarding the "nextLink" part before sending the current iteration.
Let me address that with a new iteration.
So as far as I understand you, I should add x-ms-pageable for any function that returns a list, even if we have not implemented paging for those functions? We have some functions where we have not added paging. For those I excluded x-ms-pageable. Should I add for them also?

annatisch commented:
Thanks @gamitra - yes that's correct. Regardless of whether the service will actually page the response, using the x-ms-pageable extension will result in the cleanest, most consistent developer experience in the generated SDKs.

gamitra commented:
@annatisch , Thank You. I will add the x-ms-pageable for all the APIs except "/providers/Microsoft.ADHybridHealthService/services/{serviceName}/exporterrors/counts" because the output for that API is not in the form of "value":[], which is breaking Autorest validation.I am sending a new iteration soon.

gamitra committed:
Updating OperationList.json

gamitra committed:
Update swagger spec to add x-ms-pageable for all list operations

gamitra commented:
Thanks @annatisch for helping me with this process.All changes are in and ready for review.

gamitra committed:
Updating swagger spec

gamitra committed:
Reverting updates to check if build break is fixed

gamitra committed:
Incremental update to fix build related issue

gamitra committed:
Incremental Fix to identify build break issue

gamitra committed:
Incremental change to identify build break

gamitra committed:
Address review comments

gamitra committed:
Addressing review comments

gamitra committed:
Update Datafreshness.json

gamitra committed:
Fixing break

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update Datafreshness.json

gamitra committed:
Update Datafreshness.json

gamitra committed:
Last known good

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra commented:
@annatisch , I have updated the PR addressing all the review comments.Changes are:
1.  All _list apis have nextLink
2. Updated operationId using suggested naming convention
3. Created separate group for servicemembers.
4. Renamed all "post" to "add", including exmaple files.

Please let me know if you have any more feedback.

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+        "tags": [ "Configuration" ],
+        "description": "Gets the details of a tenant onboarded to Azure Active Directory Connect Health.",
+        "x-ms-examples": {
+          "configuration_get": {
```
Wondering if this should be "tenant_get"? Not saying it necessarily *should* be - you better know the terminology associated with your service - just wondering :)

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+            "$ref": "./examples/PatchTenant.json"
+          }
+        },
+        "operationId": "configuration_update",
```
Same question regarding "configuration" vs "tenant"

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+      }
+    },
+    "/providers/Microsoft.ADHybridHealthService/configuration": {
+      "post": {
```
This post/add operation has no payload - is that the correct behaviour?

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+            "type": "string"
+          },
+          {
+            "name": "nextLink",
```
nextLink should not need to be exposed here.

gamitra replied:
Yes you are right. I have removed it.

annatisch reviewed (changes requested):
Thanks for the updates!
In addition to the comments in the review - there's some oepration capitalization that should be corrected:

- "servicemembers_foo" -> "serviceMembers_foo"
- "addsservices_foo" -> "adDsServices_foo" or "addsServices_foo" (your call)
- "addomainservicemembers_foo" -> "adDomainServiceMembers_foo"
- "addsservicemembers_foo" -> "adDsServiceMembers_foo" or "addsServiceMembers_foo" (up to you)

gamitra commented:
@annatisch, I am naming the operation based on the routeprefix, like servicemembers, addsservices,etc. My question is if I change that pattern then is that foline? Our code is already there and  changing that to match this capitalization pattern will be breaking change for us.

annatisch commented:
@gamitra - do you already have an SDK published using these naming conventions?

gamitra commented:
@annatisch, Yes we have our service public since 2014. We were missing a swagger spec so this PR.
Hence changing these names will be breaking change for us.

annatisch commented:
@gamitra - sorry I should have clarified - I'm aware that the service is public - but changing these operation IDs will not require a change in the service. They only have an affect on the generated SDK clients.
Therefore:
- Do you have a published SDK (e.g. C#) using these operation IDs? If so this is problematic, as there should not be a published SDK without a published swagger spec.
- Are there any other existing swagger specs that use these operation IDs? Or another published spec you plan to merge with this one that uses these operation IDs? This is really the only circumstances under which the operation IDs should not be changed.

gamitra commented:
@annatisch, thanks for explaining. These two scenarios don't apply to us. I will go ahead an implement the casing comments that you suggested and send a new iteration.
Let me know if there are any more suggestions, so that I can combine them all and send the next iteration.

gamitra commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+        "tags": [ "Configuration" ],
+        "description": "Onboards a tenant in Azure Active Directory Connect Health.",
+        "x-ms-examples": {
+          "configuration_add": {
```
Same reason as above, tenant_add might bring a wrong message to users. We dont add tenants as per the traditional sense. We add configurations based on tenantId and that is a tenant for our service.

gamitra commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+            "type": "string"
+          },
+          {
+            "name": "nextPartitionKey",
```
I need your suggestion here. The nextRowKey and nextPartitionKey were introduced to allow optimization from UI calls to our APIs. But that was never carried forward. I checked the code again and I see this is marked as mandate parameter but passed as string.Empty.
How to handle this scenario in swagger spec?

gamitra commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+        "tags": [ "Adds" ],
+        "description": "Gets the forest summary for a given Active Directory Domain Service, that is onboarded to Azure Active Directory Connect Health.",
+        "x-ms-examples": {
+          "addsservices_getforestSummary": {
```
Done.

gamitra commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+            "$ref": "./examples/ForestSummary.json"
+          }
+        },
+        "operationId": "addsservices_getforestSummary",
```
Done

gamitra commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+            "type": "string"
+          },
+          {
+            "name": "nextPartitionKey",
```
Need your suggestion here. See my comment above

gamitra commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5628 @@
+            "$ref": "./examples/Alerts.json"
+          }
+        },
+        "operationId": "addsservices_listServerAlerts",
```
Done

gamitra committed:
Addressing review comments

gamitra commented:
@annatisch ,

I have addressed all your feedback comments, except the one related to "nextPartitionKey".

I need your suggestion here. The nextRowKey and nextPartitionKey were introduced to allow optimization from UI calls to our APIs. But that was never carried forward. I checked the code again and I see this is marked as mandate parameter but passed as string.Empty.
How to handle this scenario in swagger spec?

gamitra commented:
@annatisch , 
For point 1: You are absolutely right. 

For point 2: These are mandatory parameters so yes user has to supply string.Empty.
"If these values must always be sent from the client as an empty string - then it would be best to set the type to an enum with a single possible value of [""]. This will be generated as a constant in the clients and so hidden from the user."

Do you have some example links using which I can achieve this? Looks like this is what I want to achieve here.

annatisch commented:
Sure @gamitra,
An example of a single value enum can be seen here:
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/iothub/resource-manager/Microsoft.Devices/stable/2017-07-01/iothub.json#L1923

The documentation for x-ms-enum options is here:
https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-enum

gamitra commented:
@annatisch  Thank You.
Does this look fine:
            {"name": "nextPartitionKey",
            "in": "query",
            "description": "The next partition key to query for.",
            "required": true,
            "enum": [ "" ],
            "type": "string",
            "x-ms-enum": {
              "name": "nextPartitionKey",
              "modelAsString": "true"}}

annatisch commented:
@gamitra - errm yes I *think* it looks okay - though to be honest I wont know for sure until you commit and I can review the generated Python client. I can read Python much better than I can read swagger ;)

gamitra committed:
Updating nextPartitionKey and nextRowKey based on review feedback

gamitra commented:
@annatisch , Thank You :). I sent another iteration. :)

gamitra commented:
@annatisch , I dont think the enum suggestion is working. I read a post online which says to use enum for empty string the parameter type should be boolean.(https://stackoverflow.com/questions/38402648/how-to-pass-empty-value-to-a-query-param-in-swagger)
Any thoughts?

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra commented:
@annatisch , added thechanges based on your suggestion. Thanks for helping :)

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5641 @@
+        "tags": [ "Adds" ],
+        "description": "Gets the details of the Active Directory Domain servers, for a given Active Directory Domain Service, that are onboarded to Azure Active Directory Connect Health.",
+        "x-ms-examples": {
+          "addsServicemembers_list": {
```
@gamitra - Last capitalization correction => "addsServiceMembers_foo"

gamitra replied:
Yeah, looks like this was missed. I will send an update right away

gamitra committed:
Update ADHybridHealthService.json

annatisch commented:
@gamitra - could you take a look at the model validation errors in the CI?
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/374494959

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Adding new example

gamitra committed:
Update Tenant.json

gamitra committed:
Update ReplicationSummary.json

gamitra committed:
Update DeleteService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra commented:
@annatisch , fixing them. Will update you once done.

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update Services.json

gamitra committed:
Add files via upload

gamitra committed:
Add files via upload

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update AddService.json

gamitra committed:
Add files via upload

gamitra committed:
Add files via upload

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Add files via upload

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Add files via upload

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Add files via upload

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update Server_Alerts.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update ServiceConfiguration.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Add files via upload

gamitra committed:
Update AddServiceMembers.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update AddServiceMembers.json

gamitra committed:
Update ADHybridHealthService.json

gamitra committed:
Update AddsServiceMembers.json

gamitra committed:
Update AddServiceMembers.json

gamitra committed:
Update AddomainServiceMembers.json

gamitra committed:
Update ServiceMembers.json

gamitra committed:
Update ServiceMember.json

gamitra committed:
Update AddomainServiceMembers.json

gamitra commented:
@annatisch , I resolved all the issues for .12. I till see . 9, .10 and .15 failing but the errors does not look related to me. So I am not sure if I am supposed to fix them. If yes, I need help understanding how to fix them ,because the errors are quite cryptic .

Please let me know if there is any thing pending on me. Thank You

annatisch commented:
Thanks @gamitra! The build looks good now. I'll take a last look over today but we should be good to go :)

gamitra commented:
@annatisch , Thanks for your help. Looking forward to get this in :)

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5772 @@
+        "operationId": "services_listUserBadPasswordReport",
+        "parameters": [
+          {
+            "name": "serv
```
Looks like this should be x-ms-pageable

gamitra replied:
I have it as pageable. See line 2039. Am I missing something here?

annatisch replied:
@gamitra - sorry that comment somehow ended up on completely the wrong line - let me find the correct one...

annatisch commented on specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json:
```diff
@@ -0,0 +1,5772 @@
+            "$ref": "./examples/ReplicationSummary.json"
+          }
+        },
+        "operationId": "addsServices_getReplicationSummary",
```
Should this be listReplicationSummary and use x-ms-pageable?

gamitra replied:
You are right.Changing it.

annatisch reviewed (commented):
@gamitra  - last couple of questions

gamitra committed:
Update ADHybridHealthService.json

annatisch commented:
@gamitra - sorry for some reason the comment ended up referencing the wrong line number if the spec - the operation I meant to tag was "serviceMembers_listConnectors":
https://github.com/gamitra/azure-rest-api-specs/blob/adhybridhhealth_swagger/specification/adhybridhealthservice/resource-manager/Microsoft.ADHybridHealthService/stable/2014-01-01/ADHybridHealthService.json#L2303

gamitra committed:
Update ADHybridHealthService.json

gamitra commented:
@annatisch  sent a new iteration with the changes

gamitra commented:
@annatisch,

Thanks a lot for reviewing and guiding through this PR. :)

annatisch commented:
@gamitra - you're welcome - thanks for your patience :)
